### PR TITLE
Support for the Clang + Windows build scenario

### PR DESCRIPTION
**Pull request by vladipus**
This also contains various fixes for CMake pipeline and tests.

### CODE REVIEW

**Comment by sezero** [comment]
Squashing the commits and rebasing to current git would help reviewing easier:
current situation is somewhat a mess.

Why are you rejecting clang for `-Wl,--no-undefined`?  The only case that would
be necessary, as I remember, is when sanitize options are in effect.

**Comment by vladipus** [comment]
> Squashing the commits and rebasing to current git would help reviewing easier:
> current situation is somewhat a mess.

I'm using a "Files changed" tab here on GitHub, it summarizes ALL of the changes together.

> Why are you rejecting clang for `-Wl,--no-undefined`? The only case that would
> be necessary, as I remember, is when sanitize options are in effect.
These are for EXTRA_LDFLAGS linker flags. Otherwise, a warning it emitted in the console. At least for Clang, not sure about the GNU family.

**Comment by sezero** [comment]
> > Squashing the commits and rebasing to current git would help reviewing easier:
> > current situation is somewhat a mess.
> 
> I'm using a "Files changed" tab here on GitHub, it summarizes ALL of the changes together.

That "Files changed" tab may be useful for quick viewing, but if the PR is OK'ed
and get merged, it will be a mess in the git commit history.

Also, one can review the changes by retrieving the PR patch by wget like I do
(e.g. `wget https://github.com/libsdl-org/SDL/pull/4062.patch`) in which case it
will be difficult to see things through

> > Why are you rejecting clang for `-Wl,--no-undefined`? The only case that would
> > be necessary, as I remember, is when sanitize options are in effect.
> 
> These are for EXTRA_LDFLAGS linker flags. Otherwise, a warning it emitted
> in the console. At least for Clang, not sure about the GNU family.

I'm running on Linux (and also build for Mac) and never seen a warning for that

**Comment by slouken** [comment]
It looks like there are a lot of changes squished together and not a lot of explanation about what is happening here.

Please create separate pull requests or patches with specific changes.

Thanks!

**Comment by icculus** [comment]
(Also, we're all learning the proper procedures for pull requests still, so if you need help untangling this, let me know and we'll all learn together.)

**Comment by slouken** [comment]
Also, you can always attach a patch and we can merge it manually the old way, if that's easier.

**Comment by vladipus** [comment]
> It looks like there are a lot of changes squished together and not a lot of explanation about what is happening here.
> 
> Please create separate pull requests or patches with specific changes.
> 
> Thanks!
The feature list as of commits is as follows:
* Maximum Clang + Windows scenario compatibility. …
* Unifont now bundled within tests for more portability.
* Correct SSE behavior for Clang on Windows. …

This patch is mainly for Clang compatibility, less warnings, etc. The intrinsics are also (finally) supported on Clang + Windows.

Ok, I think I'll make the patch then.

**Comment by vladipus** [comment]
> I'm running on Linux (and also build for Mac) and never seen a warning for that

Try running with Clang on Windows.

**Comment by sezero** [comment]
> > I'm running on Linux (and also build for Mac) and never seen a warning for that
> 
> Try running with Clang on Windows.

If you'd be so kind to post the warning message here, we'd be all aware
of the problems.  If windows is a problem with clang, but not other targets,
then I'd suggest adjusting the restrictions accordingly.

**Comment by vladipus** [comment]
Attaching the patch. Only TXT seems to be supported.
[clang.txt](https://github.com/libsdl-org/SDL/files/5973956/clang.txt)

**Comment by vladipus** [comment]
> > Try running with Clang on Windows.
> 
> If you'd be so kind to post the warning message here, we'd be all aware
> of the problems. If windows is a problem with clang, but not other targets,
> then I'd suggest adjusting the restrictions accordingly.

Ok, here you go. It might be some kind of issue within Clang, but it triggers the following warning:

image

**Comment by vladipus** [comment]
Attaching a patch with a more specific clause (tested).
[sdl-clang-win-compat.txt](https://github.com/libsdl-org/SDL/files/5975328/sdl-clang-win-compat.txt)

**Comment by vladipus** [comment]
Attaching a patch with a more specific clause (tested).
[sdl-clang-win-compat.txt](https://github.com/libsdl-org/SDL/files/5975329/sdl-clang-win-compat.txt)

**Comment by vladipus** [comment]
Hello? Is there anything I can do to help apply the patch?

**Comment by slouken** [comment]
Some comments:
1. Do you need to check the clang version before enabling intrinsics?
2. Why did you need to bump up the required CMake version?
3. Why did you include a font file?
4. Your patch does not apply cleanly to the latest version of SDL in git
5. It would be helpful if you separate out the patches by logical change so they can more easily be sorted out in the future.

Thanks!

File: src/cpuinfo/SDL_cpuinfo.c [hunk 1]
```diff
@@ -289,7 +289,7 @@ CPU_calcCPUIDFeatures(void)
                 /* Check to make sure we can call xgetbv */
                 if (c & 0x08000000) {
                     /* Call xgetbv to see if YMM (etc) register state is saved */
-#if defined(__GNUC__) && (defined(__i386__) || defined(__x86_64__))
+#if (defined(__GNUC__) || defined(__clang__)) && (defined(i386) || defined(__x86_64__))
```
Comment (sezero) [inline src/cpuinfo/SDL_cpuinfo.c:292]: Please don't change `__i386__` back to `i386`

Comment (vladipus) [reply src/cpuinfo/SDL_cpuinfo.c:292]: Did this only to resolve the conflicts. I guess now the supplied patch is also deprecated.

**Comment by vladipus** [comment]
> Some comments:
> 
> 1. Do you need to check the clang version before enabling intrinsics?
I'm testing with the latest official build available which is 11.0.1. Not sure about the former versions really.
> 2. Why did you need to bump up the required CMake version?
Because the main CMakeLists.txt is of that version and there is no practical reason not to. The OpenGL module checks are also not guaranteed to work in that same way on previous versions.
> 3. Why did you include a font file?
The sample doesn't work otherwise. This change makes it work out-of-the-box, with no additional work.
> 4. Your patch does not apply cleanly to the latest version of SDL in git
Because it was made week+ ago and you remained silent. You should create one now yourself and apply it manually.
> 5. It would be helpful if you separate out the patches by logical change so they can more easily be sorted out in the future.
That's true indeed. I've made the commits in my branch appropriately though...
> Thanks!
Thank you as well =)

**Comment by sezero** [comment]
Question about the changes like the following:
```
-#if defined(__GNUC__)
+#if (defined(__GNUC__) || defined(__clang__))
```
Does windows version of clang not identify itself
as gcc for compatibility?

**Comment by vladipus** [comment]
> Question about the changes like the following:
> 
> ```
> -#if defined(__GNUC__)
> +#if (defined(__GNUC__) || defined(__clang__))
> ```
> 
> Does windows version of clang not identify itself
> as gcc for compatibility?

The funny thing is that it doesn't. It's running in a weird mode, with MSVC (!) defined. And yes, I'm talking about official Clang distribution. 

There's even a separate command-line interface available, as it's detected by the latest CMake and is called Clang flavor. That doesn't relate to our case, just mentioned to get you aware of. Another thing is Clang's asm notation. It's different from Visual's and is actually GCC one, even if MSVC is defined. 

In the recent months I've dealt with all sorts of problems with this built scenario even for my own projects and other popular libs. If you really want to dig into this, you have to install the necessary stack yourselves and try to compile before and after my changes. I'm pretty sure you'll recreate all of my steps yourself, or still miss some important functionality like intrinsics, simd stuff. 

Cheers.

**Comment by sezero** [comment]
I merged #4145 which fixes cpuid code for clang-cl. Please see #4145,
which states: `The clang-cl compiler defines __llvm__ but not __GNUC__`
If you want to revise your patch, do so.  For convenience, I attached
the up-to-date version of your patchset, squashed into a single commit,
which can be merged using `git am`.

[4062_patch.txt](https://github.com/libsdl-org/SDL/files/6075996/4062_patch.txt)

**Comment by vladipus** [comment]
#4145 change is (was) a part of my original PR as you've might noticed.

I haven't made any changes since, so the present diff should be fine. Merge it! Good stuff! ;)

**Comment by sezero** [comment]
I broke the patch into pieces, and applied 10 parts of it as the following:
https://github.com/libsdl-org/SDL/commit/c1c4f274c34b5281c83191bd765c50e30312799b
https://github.com/libsdl-org/SDL/commit/e228d15261e624e9b91f578bbe20f43f3d70fddf
https://github.com/libsdl-org/SDL/commit/0ed17131288015c3fa050888ca1432655c629db7
https://github.com/libsdl-org/SDL/commit/dd5aff308aa45bcd4dcf2df253c92faf084b7a9c
https://github.com/libsdl-org/SDL/commit/e7abf96801bc17c56d05308adc894001bfa8ab63
https://github.com/libsdl-org/SDL/commit/a52d0db5328d89ba27be786b4fc81063f203db46
https://github.com/libsdl-org/SDL/commit/81cdb1bb0de4762a69d4339bc5fb278e0c3cc9ef
https://github.com/libsdl-org/SDL/commit/05cc9e37649a6e522a98a810bd5865e3bfcd546d
https://github.com/libsdl-org/SDL/commit/3880174a9f04dcb9fe67c20b2186070085023c7e
https://github.com/libsdl-org/SDL/commit/34ecd71e80b651c9840f60d91be60e30305fdc6f

The rest, I am attaching here as a patch file: 
[4062_patch.txt](https://github.com/libsdl-org/SDL/files/6084877/4062_patch.txt)

@slouken: The not-yet-applied parts in the patch attached above
is what I don't know or not sure about.  Please show your love &
care for it.

**Comment by sezero** [comment]
@vladipus: Please test git main, and see the un-applied parts of your patchset in the attachment above. I suggest that you close this PR and create a fresh one if there are changes needed.

**Comment by vladipus** [comment]
Roger that. I'm going in!

**Comment by vladipus** [comment]
Your changes seem to miss the prefetch polyfill. 
This won't compile, giving:
image
Merge the whole branch please. I've resolved the actual conflicts.

**Comment by sezero** [comment]
As I did say above:  see the un-applied parts of your patchset in the attachment
above [...]  I suggest that you close this PR and create a fresh one:  The new PR
should be rebased to main git and must have only the necessary commits, with
explanations as to why each is needed.

**Comment by vladipus** [comment]
The commits in question are already broken-up, since you've applied some of the patches already and those changes were actually a part of the features (fixes).
It makes little sense that you've applied them partially.
Let this PR stay as a reminder then.

**Comment by sezero** [comment]
I'm pinging @slouken again for it: see https://github.com/libsdl-org/SDL/pull/4062#issuecomment-790717810

**Comment by sezero** [comment]
@slouken: Possible LICENSE issue about commit https://github.com/libsdl-org/SDL/commit/81cdb1bb0de4762a69d4339bc5fb278e0c3cc9ef
which adds precompiled unifont-9.0.02.hex under tests directory.
unifont-9.0.02 COPYING file has this notice:
```
License for font files:

     All glyphs are released under the GNU General Public License
     (GPL) version 2 or (at your option) a later version, with the
     GNU font embedding exception:

          ** GPL v2.0 license with font embedding exception:

          As a special exception, if you create a document which
          uses this font, and embed this font or unaltered portions
          of this font into the document, this font does not by
          itself cause the resulting document to be covered by
          the GNU General Public License. This exception does not
          however invalidate any other reasons why the document
          might be covered by the GNU General Public License.
          If you modify this font, you may extend this exception
          to your version of the font, but you are not obligated
          to do so. If you do not wish to do so, delete this
          exception statement from your version.
```
If this is an issue, we should revert that commit.

**Comment by vladipus** [comment]
> @slouken: Possible LICENSE issue about commit [81cdb1b](https://github.com/libsdl-org/SDL/commit/81cdb1bb0de4762a69d4339bc5fb278e0c3cc9ef)
> which adds precompiled unifont-9.0.02.hex under tests directory.
> unifont-9.0.02 COPYING file has this notice:
> 
> ```
> License for font files:
> 
>      All glyphs are released under the GNU General Public License
>      (GPL) version 2 or (at your option) a later version, with the
>      GNU font embedding exception:
> 
>           ** GPL v2.0 license with font embedding exception:
> 
>           As a special exception, if you create a document which
>           uses this font, and embed this font or unaltered portions
>           of this font into the document, this font does not by
>           itself cause the resulting document to be covered by
>           the GNU General Public License. This exception does not
>           however invalidate any other reasons why the document
>           might be covered by the GNU General Public License.
>           If you modify this font, you may extend this exception
>           to your version of the font, but you are not obligated
>           to do so. If you do not wish to do so, delete this
>           exception statement from your version.
> ```
> 
> If this is an issue, we should revert that commit.

Nice, but that should not really be an issue. The hex version of the font is not modified at all and is taken from the official source.

"The GPL font exception clause (or GPL+FE, for short) is an optional clause that can be added to the GNU General Public License (GNU GPL) permitting digital fonts shared with that license to be embedded within a digital document file without requiring the document itself to also be shared with GPL"

**Comment by sezero** [comment]
> Nice find, but that should not really be an issue.

I'm not a lawyer, and it was my oversight applying that patch without
checking license first. Maintainers @slouken and @icculus should decide.

**Comment by vladipus** [comment]
> > Nice find, but that should not really be an issue.
> 
> I'm not a lawyer, and it was my oversight applying that patch without
> checking license first. Maintainers @slouken and @icculus should decide.

I'd better contact the authors of the font itself. BRB.

**Comment by icculus** [comment]
The font absolutely can't go in, sorry.

(Who licenses a font under the GPL anyhow?!)

**Comment by sezero** [comment]
> The font absolutely can't go in, sorry.

Reverted the commit: https://github.com/libsdl-org/SDL/commit/9d46bd25c7db0fcfaf70ac7d33b4058648e057ab

**Comment by vladipus** [comment]
> The font absolutely can't go in, sorry.
Ok. Can we just download the font vis CMake during the build like:
```cmake
file(DOWNLOAD https://github.com/toolworks/unifont.hex/raw/main/unifont-9.0.02.hex ...)
```
The command is pretty standard and should be available.

> (Who licenses a font under the GPL anyhow?!)

=)

**Comment by sezero** [comment]
> > The font absolutely can't go in, sorry.
> 
> Ok. Can we just download the font vis CMake during the build like:

I'd say don't push it

**Comment by vladipus** [comment]
> > > The font absolutely can't go in, sorry.
> > 
> > 
> > Ok. Can we just download the font vis CMake during the build like:
> 
> I'd say don't push it

The original intention for this font be available build-time is to be able to test all of the cases out-of-the-box.
Otherwise a user has to dig into the sources and research the whole unifont thingy.

**Comment by vladipus** [comment]
@sezero @slouken 
The issue is resolved now. The newer version of Unifont is licensed under OFL. I'll update the file and the sources.

**Comment by vladipus** [comment]
image

Done, tested.

**Comment by vladipus** [comment]
How's it going? Have you other questions?

**Comment by vladipus** [comment]
Closing in favor of:
https://github.com/libsdl-org/SDL/pull/4269
https://github.com/libsdl-org/SDL/pull/4268

**Comment by gvanem** [comment]
Good to see clang-cl is supported better. But still I'm getting these warnings:
```
video/windows/SDL_windowsmessagebox.c(745,89): warning: cast between incompatible calling conventions 
'cdecl' and 'stdcall'; calls through this pointer may abort at runtime [-Wcast-calling-convention]
  ...= DialogBoxIndirectParam(NULL, (DLGTEMPLATE*)dialog->lpDialog, ParentWindow, (DLGPROC)MessageBoxDialogProc, (LPA...
                                                                                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

A `static INT_PTR WINAPI MessageBoxDialogProc(HWND hDlg, UINT iMessage, WPARAM wParam, LPARAM lParam)` fixes this.

And:
```c
atomic/SDL_spinlock.c(182,5): warning: implicitly declaring library function '_ReadWriteBarrier' with type 'void (void)'
      [-Wimplicit-function-declaration]
```
since `<intrin.h>` is not included (but `<xmmintrin.h>` is).

**Comment by sezero** [comment]
> ```
> video/windows/SDL_windowsmessagebox.c(745,89): warning: cast between incompatible calling conventions 
> 'cdecl' and 'stdcall'; calls through this pointer may abort at runtime [-Wcast-calling-convention]
>   ...= DialogBoxIndirectParam(NULL, (DLGTEMPLATE*)dialog->lpDialog, ParentWindow, (DLGPROC)MessageBoxDialogProc, (LPA...
>                                                                                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> ```

This one should be fixed by https://github.com/libsdl-org/SDL/commit/227021b6470d07f86e209d89f176ec8e6d952a17

**Comment by sezero** [comment]
> ```c
> atomic/SDL_spinlock.c(182,5): warning: implicitly declaring library function '_ReadWriteBarrier' with type 'void (void)'
>       [-Wimplicit-function-declaration]
> ```
> 
> since `<intrin.h>` is not included (but `<xmmintrin.h>` is).

Made this is a new issue - #4346

**Pull request closed without merging**